cdc/cdc.cc

```diff
@@ -63,6 +65,168 @@ using namespace std::chrono_literals;
 
 static logging::logger cdc_log("cdc");
 
+namespace cdc {
+static schema_ptr create_log_schema(const schema&, std::optional<utils::UUID> = {});
+static schema_ptr create_stream_description_table_schema(const schema&, std::optional<utils::UUID> = {});
+static future<> populate_desc(db_context ctx, const schema& s);
+}
+
+class cdc::cdc_service::impl : service::migration_listener::empty_listener {
+    db_context _ctxt;
+public:
+    impl(db_context ctxt)
+        : _ctxt(std::move(ctxt))
+    {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.register_listener(this);
+        }
+    }
+    ~impl() {
+        if (engine().cpu_id() == 0) {
+            _ctxt._migration_manager.unregister_listener(this);
+        }
+    }
+
+    void on_before_create_column_family(const schema& schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        if (schema.cdc_options().enabled()) {
+            auto& db = _ctxt._proxy.get_db().local();
+            auto logname = log_name(schema.cf_name());
+            if (!db.has_schema(schema.ks_name(), logname)) {
+                // in seastar thread
+                auto log_schema = create_log_schema(schema);
+                auto stream_desc_schema = create_stream_description_table_schema(schema);
+                auto& keyspace = db.find_keyspace(schema.ks_name());
+
+                auto log_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), log_schema, timestamp);
+                auto stream_mut = db::schema_tables::make_create_table_mutations(keyspace.metadata(), stream_desc_schema, timestamp);
+
+                mutations.insert(mutations.end(), std::make_move_iterator(log_mut.begin()), std::make_move_iterator(log_mut.end()));
+                mutations.insert(mutations.end(), std::make_move_iterator(stream_mut.begin()), std::make_move_iterator(stream_mut.end()));
+            }
+        }
+    }
+
+    void on_before_update_column_family(const schema& new_schema, const schema& old_schema, std::vector<mutation>& mutations, api::timestamp_type timestamp) override {
+        bool is_cdc = new_schema.cdc_options().enabled();
+        bool was_cdc = old_schema.cdc_options().enabled();
+
+        if (is_cdc != was_cdc || is_cdc) {
+            auto logname = log_name(old_schema.cf_name());
+            auto descname = desc_name(old_schema.cf_name());
+            auto& db = _ctxt._proxy.get_db().local();
+            auto& keyspace = db.find_keyspace(old_schema.ks_name());
+            auto log_schema = was_cdc ? db.find_column_family(old_schema.ks_name(), logname).schema() : nullptr;
+            auto stream_desc_schema = was_cdc ? db.find_column_family(old_schema.ks_name(), descname).schema() : nullptr;
+
+            if (!is_cdc) {
+                auto log_mut = db::schema_tables::make_drop_table_mutations(keyspace.metadata(), log_schema, timestamp);
```

Comment:
I know it's an invariant that
`cdc was enabled => log schema and desc schema exist`
but what if there's some subtle race or just bug where we break this invariant (maybe just for a short period of time)?

I'd suggest to make a check here that log_schema != nullptr. (i.e., `find_column_family` succeeded). Maybe not an assert, but if + throw runtime_error. Or simply log it with warn level (in `drop` I'd go for log, in `update` I'd go for an error probably).

Reply:
log_schema etc are always created in the same batch of mutations as base schema. We cannot race two schema updates at the same time (uses semaphore lock), and we are not notified until all changes are done (still inside lock). 
If log_schema was null here with was_cdc=true we should honestly either crash or assert, because there is no valid way we can reach that state _unless_ we had a crash in a run before, between two "db.add_column_family_and_make_directory" calls. 
But sure, an assert might do.

Reply:
How can log_schema be null? It was defined above to not null if was_cdc... If the find_column_family() above failed, we would get an exception here, not a null.

Reply:
> If the find_column_family() above failed, we would get an exception here, not a null

For some reason I was convinced it would return `nullptr`. Sorry for introducing the confusion